library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
-    def factory(module):
-        type = module.params.get('connection')
+        if remote:
+            software = self.handle_remote_source(link)
+            self._values['software_name'] = software
+            if link_md5:
+                image_md5 = self.handle_remote_source(link_md5)
+                self._values['imgmd5_name'] = image_md5
 
-        if type == "rest":
-            module.fail_json(msg='The REST connection is currently not supported')
-        elif type == "soap":
-            if not BIGSUDS_AVAILABLE:
-                raise Exception("The python bigsuds module is required")
-            return BigIpSoapApi(check_mode=module.check_mode, **module.params)
+        else:
+            software = os.path.split(link)[1]
+            self._values['software_name'] = software
 
-    factory = staticmethod(factory)
+        return self._values['software']
 
+    @property
+    def hotfix(self):
+        link = self._values['hotfix']
+        link_md5 = self._values['hotfix_md5sum']
+        remote = self.remote_src
 
-class BigIpCommon(object):
-    def __init__(self, *args, **kwargs):
-        self.result = dict(changed=False, changes=dict())
+        if link is None:
+            return None
 
-        self.current = dict()
+        if remote:
+            hotfix = self.handle_remote_source(link)
+            self._values['hotfix_name'] = hotfix
+            if link_md5:
+                hotfix_md5 = self.handle_remote_source(link_md5)
+                self._values['hfmd5_name'] = hotfix_md5
 
-        if kwargs['hotfix']:
-            kwargs['photfix'] = self.iso_info(kwargs['hotfix'])
+        else:
+            hotfix = os.path.split(link)[1]
+            self._values['hotfix_name'] = hotfix
 
-        if kwargs['software']:
-            kwargs['psoftware'] = self.iso_info(kwargs['software'])
+        return self._values['hotfix']
 
-        self.params = kwargs
+    @property
+    def options(self):
+        state = self._values['state']
+        create = self._values['create_volume']
+        opt = list()
 
-    def _find_iso_content(self, iso):
-        paths = ['/METADATA.XML', 'metadata.xml']
+        if create:
+            opt.append({'create-volume': True})
+        if state == 'activated':
+            opt.append({'reboot': True})
 
-        for path in paths:
-            try:
-                content = iso.record(path).content
-                return content
-            except KeyError:
-                pass
-        raise Exception(
-            "Unable to find metadata file in ISO. Please file a bug"
-        )
+        return opt
 
-    def iso_info(self, iso):
-        result = dict(
-            product=None,
-            version=None,
-            build=None
-        )
+    @property
+    def volume(self):
+        state = self._values['state']
+        volume = self._values['volume']
+        reuse = self._values['reuse_inactive_volume']
+
+        if state == 'activated' or state == 'installed':
```

Comment:
avoid using a large if block like this. instead do

if state in ['absent', 'present']:
    return self._values['volume']

and then backspace everything that was formerly under the if statement.